analyzer/validatorbalances/validatorbalances.go

```diff
@@ -0,0 +1,109 @@
+package evmtokens
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/oasisprotocol/nexus/analyzer"
+	"github.com/oasisprotocol/nexus/analyzer/item"
+	"github.com/oasisprotocol/nexus/analyzer/queries"
+	"github.com/oasisprotocol/nexus/config"
+	staking "github.com/oasisprotocol/nexus/coreapi/v22.2.11/staking/api"
+	"github.com/oasisprotocol/nexus/log"
+	"github.com/oasisprotocol/nexus/storage"
+	"github.com/oasisprotocol/nexus/storage/oasis/nodeapi"
+)
+
+// The validator balance analyzer (1) gets a list of epochs to download validator
+// balance info for, (2) downloads that info, and (3) saves the info in the database.
+
+const (
+	//nolint:gosec // thinks this is a hardcoded credential
+	validatorBalanceAnalyzerName = "validator_balance_"
+)
+
+type processor struct {
+	source nodeapi.ConsensusApiLite
+	target storage.TargetStorage
+	logger *log.Logger
+}
+
+var _ item.ItemProcessor[*Epoch] = (*processor)(nil)
+
+func NewAnalyzer(
+	cfg config.ItemBasedAnalyzerConfig,
+	sourceClient nodeapi.ConsensusApiLite,
+	target storage.TargetStorage,
+	logger *log.Logger,
+) (analyzer.Analyzer, error) {
+	logger = logger.With("analyzer", validatorBalanceAnalyzerName)
+	p := &processor{
+		source: sourceClient,
+		target: target,
+		logger: logger,
+	}
+
+	return item.NewAnalyzer[*Epoch](
+		validatorBalanceAnalyzerName,
+		cfg,
+		p,
+		target,
+		logger,
+	)
+}
+
+type Epoch struct {
+	Epoch       uint64
+	StartHeight int64
+}
+
+func (p *processor) GetItems(ctx context.Context, limit uint64) ([]*Epoch, error) {
+	var epochs []*Epoch
+	rows, err := p.target.Query(ctx, queries.ConsensusValidatorEpochs, limit)
+	if err != nil {
+		return nil, fmt.Errorf("querying epochs for validator history: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var epoch Epoch
+		if err = rows.Scan(
+			&epoch.Epoch,
+			&epoch.StartHeight,
+		); err != nil {
+			return nil, fmt.Errorf("scanning epoch: %w", err)
+		}
+		epochs = append(epochs, &epoch)
+	}
+	return epochs, nil
+}
+
+func (p *processor) ProcessItem(ctx context.Context, batch *storage.QueryBatch, epoch *Epoch) error {
+	p.logger.Info("downloading validator balances", "epoch", epoch)
+	validators, err := p.source.GetValidators(ctx, epoch.StartHeight)
```

Comment:
This ends up calling oasis-core `GetValidators` right? This returns just the active validators for the epoch starting at that height. 

I think we maybe don't want just these, but go over all known consensus accounts since escrow balance can change to any account? (and debonding as well)

Reply:
If we update this to track for all accounts (not just "validators" for the epoch) then we could also use this historic table for things like `"account birth/created_at"` which is also a request from the frontend team.

Unless it is not feasible to go over all accounts for every epoch, then we need to think about it a bit more how to approach the "account" stats.

Reply:
I expect that **grabbing all accounts at every epoch is untenable**. We'd probably need `StateToGenesis`? Or something else that will be similarly slow, since staking data is a huge part of `StateToGenesis` output.
To run this in real time, we'd need `StateToGenesis` to be faster than 60min (= 1 epoch). Which it is, even if only when we are somewhat careful about the node's hardware (disks). But to run this in fast-sync, even a 10-minute response would mean we need ~0.5years to process 3 years of blocks. We can get around it by using multiple nodes, but I'm not sure if it's worth it.

Also, downloading so much data would put a strain on (and possibly force us to rethink/redesign) the DB, caches, etc. It's a lot of data. Very conservatively: (1e6 accounts) * (32000 epochs) * (100 bytes per account snapshot at given epoch) = 3.2TB, but easily a few times more.

I believe we limited ourselves to just validators because it's more important that we be able to show the history of a validator (as it builds their credibility), compared to the history of a "normal" user.

But point well taken; if somebody stops being an active validator and later becomes one again, their interim history matters. Andy, how many accounts would we need to touch **if we queried every account that was ever a validator? Maybe that's a reasonable compromise.**

If this is slow to run or complicated to implement, I strongly **suggest we go with what we already have in this PR** at least for now, and thus potentially report partial histories for validators. It's much better than nothing, it's 0 additional effort, and it's launchable.

Reply:
>  if we queried every account that was ever a validator

I think this should be feasable. The validators dont flactuate a lot. I believe there are 226 such validators on mainnet (based on oasisscan, i think they track validators in a similar way to this).

Reply:
Thanks all for the feedback! Agreed that querying all accounts that were ever a validator is a good compromise. 

One potential complication - having a batch size > 1 can cause issues if a new validator appears. Eg, if we're processing epochs 1-20 in parallel but epoch 3 introduces a new validator, the worker processing epochs 4-20 won't know about the new validator.

For a given epoch N, we want to fetch info for all the validators fetched for epoch N-1, `union`-ed with the set of active validators for epoch N. This sequential dependency would require us to always run this analyzer with a batch size of 1 so we can build up the list of known validators. Compute-wise, this _should_ be feasible given that there are ~35k epochs, but I'll need to do some testing to confirm. Another caveat is that we won't have staked balance history for an account before it becomes a validator. Imo this is fine, since we won't be displaying any deep history and it won't hinder staking rewards calculations when we add that in the future.

We could instead hardcode the 226 validators and have the analyzer fetch the staked balance for each one up to some recent block, and then have the analyzer just proceed sequentially(batch_size=1) from then onwards.

Another approach would be to treat each (epoch, validator) as the item instead of just (epoch). I need to do more work to see how complicated the item accounting becomes, but maybe it'll produce a nicer solution.

Reply:
Updated to have the analyzer process the epochs sequentially from a given `start_height`, taking the union of {current validators} x {validators from prior epoch}. The start height is expected to match consensus analyzer start height, but could be adjusted for a faster reindex. This allows us to track the continuous history for each validator from the epoch they first become a validator, even if the validator drops out after.

Having the analyzer process epochs sequentially ends up not constraining performance since each epoch makes >100 node requests for the validator balances.